taichi/rhi/metal/metal_device.mm

```diff
@@ -530,6 +605,82 @@ DeviceCapabilityConfig collect_metal_device_caps(MTLDevice_id mtl_device) {
   return std::make_unique<MetalSampler>(mtl_sampler_state);
 }
 
+MetalSurface::MetalSurface(MetalDevice *device, const SurfaceConfig &config)
+    : config_(config), device_(device) {
+
+  width_ = config.width;
+  height_ = config.height;
+
+  image_format_ = BufferFormat::bgra8;
+
+  layer_ = [CAMetalLayer layer];
+  layer_.device = device->mtl_device();
+  layer_.pixelFormat = format2mtl(image_format_);
+  layer_.drawableSize = CGSizeMake(width_, height_);
+  layer_.allowsNextDrawableTimeout = NO;
+#if TARGET_OS_MAC
+  // Older versions may not have this property, so check if it exists first.
+  layer_.displaySyncEnabled = config.vsync;
+#endif
+}
+
+MetalSurface::~MetalSurface() {
+  destroy_swap_chain();
+  [layer_ release];
+}
+
+void MetalSurface::destroy_swap_chain() {
+  for (auto &alloc : swapchain_images_) {
+    device_->destroy_image(alloc.second);
+  }
+  swapchain_images_.clear();
+}
+
+StreamSemaphore MetalSurface::acquire_next_image() {
+  current_drawable_ = [layer_ nextDrawable];
+  current_swap_chain_texture_ = current_drawable_.texture;
+
+  if (swapchain_images_.count(current_swap_chain_texture_) == 0) {
+    swapchain_images_[current_swap_chain_texture_] =
+        device_->import_mtl_texture(current_drawable_.texture);
+    RHI_ASSERT(swapchain_images_.size() <=
+               50); // In case something goes wrong on Metal side, prevent this
+                    // map of images from growing each frame unbounded.
+  }
+  return nullptr;
+}
+
+DeviceAllocation MetalSurface::get_target_image() {
+  return swapchain_images_.at(current_swap_chain_texture_);
+}
+
+void MetalSurface::present_image(
+    const std::vector<StreamSemaphore> &wait_semaphores) {
+
+  Stream *stream = device_->get_compute_stream();
+  auto [cmd_list, res] = stream->new_command_list_unique();
+  MetalCommandList *cmd_list2 = (MetalCommandList *)cmd_list.get();
+  RHI_ASSERT(res == RhiResult::success);
+
+  [cmd_list2->finalize() presentDrawable:current_drawable_];
```

Comment:
It's probably cleaner to use the Metal API directly here instead of using the RHI version.

Reply:
Also after checking the API for a bit I think you don't need an command buffer for this. Here's an API:
https://developer.apple.com/documentation/metal/mtldrawable/1470284-present?language=objc

From the commandbuffer API it says the presentDrawable is just a convenience function wrapping this underlying `present` function